perf/bps_tree.cc

```diff
@@ -0,0 +1,464 @@
+	}
+
+	static void
+	free(void *ctx, void *extent)
+	{
+		(void)ctx;
+		(void)extent;
+	}
+};
+
+/**
+ * Each tree configuration (tree_i64, tree_s128, etc.) has a class associated.
+ * This makes it possible to create generic benchmarks using C++ templates.
+ */
+
+#define CREATE_TREE_CLASS(tree) \
+class tree { \
+public: \
+	using tree_t = tree##_t; \
+	using elem_t = tree##_elem_t; \
+	using key_t = tree##_key_t; \
+	using Allocator_Base = \
+		DummyAllocator<tree##_EXTENT_SIZE>; \
+	struct Allocator: public Allocator_Base { \
+		Allocator(size_t count) \
+		: Allocator_Base(count * sizeof(elem_t) * 2) \
+		{} \
+	}; \
+	static constexpr size_t max_count_in_leaf = \
+		BPS_TREE_##tree##_t_MAX_COUNT_IN_LEAF; \
+	static constexpr size_t max_count_in_inner = \
+		BPS_TREE_##tree##_t_MAX_COUNT_IN_INNER; \
+	static constexpr size_t height_1_max_size = max_count_in_leaf; \
+	static constexpr size_t height_2_max_size = height_1_max_size * \
+						    max_count_in_inner; \
+	static constexpr size_t height_3_max_size = height_2_max_size * \
+						    max_count_in_inner; \
+	static constexpr size_t height_4_max_size = height_3_max_size * \
+						    max_count_in_inner; \
+	static constexpr size_t height_5_max_size = height_4_max_size * \
+						    max_count_in_inner; \
+	static constexpr auto create = ::tree##_t_create; \
+	static constexpr auto build = ::tree##_t_build; \
+	static constexpr auto destroy = ::tree##_t_destroy; \
+	static constexpr auto find = ::tree##_t_find; \
+	static constexpr auto insert = ::tree##_t_insert; \
+	static constexpr auto delete_ = ::tree##_t_delete; \
+}
+
+/* The class must be created for each instantiated BPS tree to test it. */
+CREATE_TREE_CLASS(tree_i64);
+
+/**
+ * Value generators to make key-independent benchmarks.
+ */
+
+class ValueKey {
+	size_t value;
+public:
+	ValueKey(size_t value) : value(value) {}
+
+	size_t
+	operator()()
+	{
+		return value;
+	}
+};
+
+class RandomKey {
+	size_t mod;
+	std::minstd_rand rng;
+public:
+	RandomKey(size_t mod) : mod(mod) {}
+
+	size_t
+	operator()()
+	{
+		return rng() % mod;
+	}
+};
+
+class IncrementingKey {
+	size_t value;
+public:
+	IncrementingKey() : value(0) {}
+
+	size_t
+	operator()()
+	{
+		return value++;
+	}
+};
+
+class DecrementingKey {
+	size_t value;
+public:
+	DecrementingKey(size_t init) : value(init) {}
+
+	size_t
+	operator()()
+	{
+		return --value;
+	}
+};
+
+/**
+ * The benchmarks.
+ */
+
+template<class tree>
+static void
+test_build(benchmark::State &state, size_t count)
+{
+	std::vector<typename tree::elem_t> arr;
+	for (size_t i = 0; i < count; i++)
+		arr.emplace_back(i);
+
+	typename tree::Allocator allocator(count);
+	for (auto _ : state) {
+		typename tree::tree_t t;
+		tree::create(&t, 0, allocator.alloc, allocator.free,
+			     &allocator, NULL);
+		tree::build(&t, &arr.front(), count);
+		tree::destroy(&t);
+		allocator.reset();
+	}
+}
+
+/* Tree create & destroy overhead. */
+generate_benchmark_size(tree_i64, build, 0);
```

Comment:
I don't think that the destruction overhead is interesting. Only the construction is important, since it affects start up time. See also comment above.

Reply:
It meant to show the extra work is small, but even if it wouldn't it could not show it (the allocator destruction is a good example).

Removed this benchmark.